src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,356 @@
+use glib::{clone, format_size};
+use gtk::glib::SignalHandlerId;
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use gtk::{gio, glib, CompositeTemplate};
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use tdlib::enums::MessageContent;
+use tdlib::types::File;
+
+use crate::session::content::message_row::{
+    MessageBase, MessageBaseImpl, MessageIndicators, MessageLabel,
+};
+use crate::tdlib::{ChatType, Message, MessageSender};
+use crate::utils::parse_formatted_text;
+use crate::Session;
+
+use super::base::MessageBaseExt;
+
+mod imp {
+    use super::*;
+    use once_cell::sync::Lazy;
+    use std::cell::RefCell;
+
+    #[derive(Debug, Default, CompositeTemplate)]
+    #[template(resource = "/com/github/melix99/telegrand/ui/content-message-document.ui")]
+    pub(crate) struct MessageDocument {
+        pub(super) sender_color_class: RefCell<Option<String>>,
+        pub(super) bindings: RefCell<Vec<gtk::ExpressionWatch>>,
+        pub(super) status_handler_id: RefCell<Option<SignalHandlerId>>,
+        pub(super) click_handler_id: RefCell<Option<SignalHandlerId>>,
+        pub(super) message: RefCell<Option<glib::Object>>,
+        #[template_child]
+        pub(super) sender_label: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub(super) content_label: TemplateChild<MessageLabel>,
+        #[template_child]
+        pub(super) indicators: TemplateChild<MessageIndicators>,
+        #[template_child]
+        pub(super) file_name_label: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub(super) file_size_label: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub(super) file_status_button: TemplateChild<gtk::Button>,
+        #[template_child]
+        pub(super) document_box: TemplateChild<gtk::Box>,
+        #[template_child]
+        pub(super) click: TemplateChild<gtk::GestureClick>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for MessageDocument {
+        const NAME: &'static str = "ContentMessageDocument";
+        type Type = super::MessageDocument;
+        type ParentType = MessageBase;
+
+        fn class_init(klass: &mut Self::Class) {
+            Self::bind_template(klass);
+            klass.set_css_name("messagedocument");
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for MessageDocument {
+        fn constructed(&self, obj: &Self::Type) {
+            let handler_id = self
```

Comment:
In telegram desktop (and mobile too) you can't cancel download by clicking on filename
You need to directly click on cross

[Kooha-2022-09-11-15-18-16.webm](https://user-images.githubusercontent.com/63008755/189527117-6d0abb94-b9c1-4201-8c80-a29128952e10.webm)

Also I would like to make DownloadController widget that inherits button and make downloading and cancelling work

Something like this

```Rust
 fn update_document(&self, message: &Message) {
        if let MessageContent::MessageDocument(data) = message.content().0 {
            let imp = self.imp();

            // Set some labels

            let session = message.chat().session();

			imp.download_controller.set_file_and_session(data.document.document, session);

			imp.download_controller.connect_progress(clone!(@weak self as obj => move |progress| {
				obj.update_progress(progress);
			}))
        }
    }
```

Reply:
And we can use the same widget for other cancelable downloads such as images
![image](https://user-images.githubusercontent.com/63008755/189527528-1f193b4c-06fb-4d21-ace6-adc723d8d6eb.png)

(I've started work on progress indicator)

Reply:
Ok, now I understand your point. :-)
I thought we'd be able to cancel by also clicking on the message. But now it makes totally sense.

Reply:
On iOS it seems to cancel the download by also clicking on the rest of the content (tested with document and photo). I think it makes the most sense to do that so that it's coherent and the code should also be simpler.

Reply:
It would reduce complexity a lot. In this turn, let's also transform the button into an image